### ISSUE TITLE
`V2.0.0` throwing `localStorage` error in React Native.

### ISSUE THREAD

**Issue by saikat-samanta** [supabase/gotrue-js#495]
# Bug report
During `signUp` and `signInWithPassword` it's throwing `localStorage` error after `V2.0.0` update in `React Native`

## Describe the bug
I updated the package to `V2` and all the `auth` methods also. But, It's not working with React Native. It's showing some error related to `localStorage`.

**Comment by william-matz** [comment]
I believe I'm seeing the same error after upgrading to v2 w/ React Native

It's happing for me when I use `signInWithOAuth()`

```
[Unhandled promise rejection: TypeError: undefined is not an object (evaluating 'window.location.href = url')]
at node_modules/@supabase/gotrue-js/dist/main/GoTrueClient.js:723:27 in _handleProviderSignIn
at node_modules/@supabase/gotrue-js/dist/main/GoTrueClient.js:5:61 in fulfilled
```

I'm using these client options:

```typescript
export const supabase = createClient(supabaseUrl, supabaseAnonKey, {
  auth: {
    storage: AsyncStorage,
    detectSessionInUrl: false,
    autoRefreshToken: true,
    persistSession: true,
  },
});
```

supabase-js version: `2.0.4`
gotrue-js version: `2.2.1`

**Comment by william-matz** [comment]
Did some digging - looks like `window` is not undefined in react native. Details [here](https://stackoverflow.com/a/49911697/14840802). I confirmed this in my project - the window object has a bunch of stuff in it.

I was previously on v1.24.0 where this was working. [Here](https://github.com/supabase/gotrue-js/blob/4e7165e3cb971b80af2f44406e9f8b1d79f3b1c6/src/GoTrueClient.ts#L675) is the fallback that I assume handled the RN case:

```js
    try {
      // try to open on the browser
      if (isBrowser()) {
        window.location.href = url
      }
      return { provider, url, data: null, session: null, user: null, error: null }
    } catch (e) {
      // fallback to returning the URL
      if (url) return { provider, url, data: null, session: null, user: null, error: null }
      return { data: null, user: null, session: null, error: e as ApiError }
    }
```

I'm not sure why this fallback was removed from v1-v2 but I'll try to get this running and see what happens

**Comment by kangmingtay** [comment]
Hey @william-matz, just wondering, in the past, when `window.location.href` is `undefined`, how did you handle redirecting the user to the oauth provider sign-in page?

**Comment by adyatlov** [comment]
Hi @kangmingtay. Redirecting  the user is an optional functionality of the `signInWithOAuth()` function. In non-browser environments it should return the following object (https://github.com/supabase/gotrue-js/blob/v2.3.1/src/GoTrueClient.ts#L933):
```TypeScript
{ data: { provider, url }, error: null }
```
As far as I understand the React Native app then should open a browser (external or internal) that should open the login URL.

**Comment by adyatlov** [comment]
A very dirty workaround; please use it at your own risk. The following code is definitely not for production.
```TypeScript
  async function signInWithGoogle() {
    const windowBackup = window;
    window = undefined;
    const { data, error } = await supabase.auth.signInWithOAuth({
      provider: 'google',
    })
    window = windowBackup;
    ...
  }
```

**Comment by dabulashvili** [comment]
Is anyone working on this problem currently? And it there any estimation for a fix?

BTW, @adyatlov for me workaround is not working.

**Comment by bpeck81** [comment]
@adyatlov My code isn't throwing an error with your solution, but it doesnt open the browser. Should it?

**Comment by adyatlov** [comment]
@bpeck81, neither the workaround nor the fix open a browser window. Maybe it should. I'll try to extend the fix.

**Comment by bpeck81** [comment]
@adyatlov Could you?
I'm using this code and when I follow through the google authentication flow in the browser I end up at this page. Clearly I did something wrong...

Code:
```async function signInWithGoogle() {
const windowBackup = window;
window = undefined;
const { data, error } = await supabase.auth.signInWithOAuth({
    provider: 'google',
})
window = windowBackup;
console.log(data.url)
WebBrowser.openBrowserAsync(data.url)
```
Browser after authenticating:
`{"message":"No API key found in request","hint":"No 'apikey' request header or url param was found."}
`

**Comment by bpeck81** [comment]
@adyatlov I'm not sure if this is related to the bug, but how am I supposed to get the authenticated user after calling the code you gave? When I call supabase.auth.getUser() after going through the google authentication process I get null values back.

**Comment by dabulashvili** [comment]
I was able to set up google auth but with 1.x version. For 2.x version the problem is that it thinks it's a browser. They need special case for mobile, or give us provider URL and we can open it in the local browser.

@bpeck81 The flow is following: generate provider URL -> open built-in browser with provider URL and set up return URL. When process is done, you can either fetch use data with given token or receive event :

```    
    const { data: authListener } = supabase.auth.onAuthStateChange(async (event, session) => {
      console.log(`Supabase auth event: ${event}`);
      if (event === 'SIGNED_IN') {
        setUser(session.user);
      }
    });
```
I can provide full example if you need.

**Comment by kangmingtay** [comment]
@dabulashvili and @bpeck81 

@adyatlov kindly helped to make a PR to attempt to resolve this issue [here](https://github.com/supabase/gotrue-js/pull/546) (thanks again 💯 ) - it would be great if you guys can try it out and see if it solves your issue. Unfortunately, i don't have much experience with RN and i don't have the bandwidth to test this out on my end. I'm happy to merge it in once we can confirm that the PR does help to fix the problem for RN and at the same time, doesn't break things for the web folks

**Comment by dabulashvili** [comment]
I checked locally with the suggested change and it works. Thank you, waiting for merge 🤞

**Comment by dabulashvili** [comment]
``` jsx
import AsyncStorage from '@react-native-async-storage/async-storage';
import * as WebBrowser from 'expo-web-browser';
import qs from 'qs';
import { useContext } from 'react';
import { KeyboardAvoidingView, Text, TouchableOpacity, View } from 'react-native';

import { supabase } from '../../../lib/initSupabase';
import { UserContext } from '../../components/UserContext';
import { styles } from './styles';

WebBrowser.maybeCompleteAuthSession();

export const GoogleSignInButton = ({ setLoading }) => {
  const { setUser } = useContext(UserContext);

  // TODO use expo url locally and set up value for prod
  const returnUri = 'exp://192.168.0.17:19000';

  async function loginWithGoogle() {
    setLoading(true);
    try {
      const { data: providerData, error: providerError } = await supabase.auth.signInWithOAuth({
        provider: 'google',
        options: {
          redirectTo: returnUri,
        },
      });

      if (providerError) {
        console.error('[signIn] Error while creating provider', providerError);
        return null;
      }

      let result = await WebBrowser.openAuthSessionAsync(providerData.url, returnUri);

      if (!result || result.type !== 'success') {
        return null;
      }

      const params = qs.parse(result.url.replace(`${returnUri}#`, ''));
      const refreshToken = params && params['refresh_token'];

      if (!refreshToken) {
        console.error('[signIn] No refresh token', userData);
        return null;
      }

      const { data: userData, error: userError } = await supabase.auth.getUser(params['access_token']);

      if (userError) {
        console.error('[signIn] Error getting user', userError);
        return null;
      }

      const { user } = userData;
      setUser(user);
      await AsyncStorage.setItem('auth_user', JSON.stringify(user));
    } catch (e) {
      console.error('[signIn] Some problem', e);
    } finally {
      setLoading(false);
    }
  }

  const handleLogin = () => {
    loginWithGoogle();
  };

  return (
    <KeyboardAvoidingView style={styles.container} behavior="padding">
      <View style={styles.buttonContainer}>
        <TouchableOpacity onPress={handleLogin} style={styles.button}>
          <Text style={styles.buttonText}>Login</Text>
        </TouchableOpacity>
      </View>
    </KeyboardAvoidingView>
  );
};
```

Working example for supabase lib v2.x

**Comment by hf** [comment]
Fixed with #546.
